src/lib.rs

```diff
@@ -0,0 +1,452 @@
+        let cursor = self.cursor.as_mut().unwrap();
+
+        if self.num_funcs == 0 {
+            return None;
+        }
+
+        // StkSizeRecord[NumFunctions] {
+        //     uint64: Function Address
+        let addr = itry!(cursor.read_u64::<NativeEndian>());
+        //     uint64: Stack Size
+        let stack_size = itry!(cursor.read_u64::<NativeEndian>());
+        //     uint64: Record Count
+        itry!(cursor_skip(cursor, 8));
+        // } -- End of this function entry.
+
+        self.num_funcs -= 1;
+        Some(Ok(SMFunc{addr, stack_size}))
+    }
+}
+
+/// Top-level struct through which the user interfaces with the stackmap section.
+pub struct StackMapParser {
+    elf_file: elf::File,
+    num_funcs: u32,
+    num_consts: u32,
+    num_stackmaps: u32,
+}
+
+impl StackMapParser {
+    pub fn new(path: &Path) -> SMParserResult<Self> {
+        let elf_file = elf::File::open_path(path)?;
+        let num_funcs;
+        let num_consts;
+        let num_stackmaps;
+        {
+            let mut cursor = cursor_from_elf(&elf_file, 0)?;
+            Self::check_header(&mut cursor)?;
+
+            // Read in table sizes.
+            // uint32: NumFunctions
+            num_funcs = cursor.read_u32::<NativeEndian>()?;
+            // uint32: NumConstants
+            num_consts = cursor.read_u32::<NativeEndian>()?;
+            // uint32: NumRecords
+            num_stackmaps = cursor.read_u32::<NativeEndian>()?;
+        }
+
+        Ok(Self{elf_file, num_funcs, num_consts, num_stackmaps})
+    }
+
+    /// Returns the number of stackmap record entries in the stackmap section.
+    pub fn num_stackmaps(&self) -> u32 {
+        self.num_stackmaps
+    }
+
+    /// Returns the number of function entries in the stackmap section.
+    pub fn num_funcs(&self) -> u32 {
+        self.num_funcs
+    }
+
+    /// Check the stackmap header looks sane.
+    fn check_header(cursor: &mut Cursor<&Vec<u8>>) -> SMParserResult<()> {
+        // uint8: Stack Map Version
+        let version = cursor.read_u8()?;
+        if version != STACKMAP_VERSION {
+            let msg = format!("Expected stackmap format v{} but binary is v{}", STACKMAP_VERSION, version);
+            return Err(SMParserError::Other(msg));
+        }
+        // uint8: Reserved (expected to be 0)
+        let b2 = cursor.read_u8()?;
+        if b2 != 0 {
+            let msg = format!("Expected 0 in stackmap section byte 2, got {}", b2);
+            return Err(SMParserError::Other(msg));
+        }
+        // uint16: Reserved (expected to be 0)
+        let b2_3 = cursor.read_u16::<NativeEndian>()?;
+        if b2_3 != 0 {
+            let msg = format!("Expected 0 in stackmap section bytes 2 and 3, got {}", b2_3);
+            return Err(SMParserError::Other(msg));
+        }
+        Ok(())
+    }
+
+    /// Make an iterator over the stackmap record entries in the stackmap section.
+    ///
+    /// If the iterator returns an error, the iterator becomes invalid.
```

Comment:
And...? Do you mean something like "Pumping the iterator for values after an error is returned leads to undefined behaviour"?

Reply:
You will get the same error over and over I think. Our hands are tied because Rust iterators can only return `Some` or `None`, where `None` stops the iterator. So the only way to return an error is to put it inside a `Some`.

We could have an "errored" flag inside which causes an extra pump to return `None` if set?

Reply:
I strongly suggest using the sentence I wrote (or some minor variant): what you need to make clear to users is "do X and bad things happen". "Undefined behaviour" strikes fear into every right-thinking programmer.

Reply:
I agree, but we can also just prevent the UB altogether with the above suggestion.

Reply:
Yes, but a) it slows the code down b) the user's still got themselves into a sticky situation that means their code is wrong. In such cases, I am a fan of putting a sign saying "warning: minefield. btw, we're not going to tell you where the mines are" rather than trying to document the mines.

Reply:
fair enough.